doc/docs/spec/typeclasses/index.md

```diff
@@ -0,0 +1,124 @@
+# Typeclasses
+
+Typeclasses are a feature that comes from Haskell as a means of achieving overloading of sorts. In Vapor this is different,
+as the main goal to achieve with typeclasses is allowing users to customize behavior, but also in more ways than once. This
+document tries to specify how they look like and how they behave.
+
+## Syntax
+
+The typeclass support encompasses the following keywords:
+
+ * `typeclass`, used as a kind for the declaration of typeclasses.
+ * `implicit`, which signifies that a given typeclass is - by design - to be inferred for all types, if no instances are
+   defined. **Drafting node: need an opt-out mechanism.**
+ * `instance`, used to declare an instance of a typeclass.
+ * `default`, which marks the instance that is used when no instance is explicitly specified.
+ * `with`, which will ultimately be used properly for introducing all kinds of templates, but right now its usage is
+   confined to defining typeclasses, as templates are not a feature to fully land in the minimal viable product.
+
+Similarly to structures, typeclasses can be defined in the named an unnamed forms; the named one is translated to a declaration
+of an entity initialized with the unnamed form of the definition. The grammar is as follows (this duplicates some terms that
+will be defined elsewhere, but not all of them, since quoting the entire grammar of expressions here would be excessive):
+
+```
+named-typeclass-declaration ::= template-introducer `implicit`? `typeclass` identifier
+    `{` typeclass-function* `}` `;`
+typeclass-literal ::= template-introducer `implicit`? `typeclass` `{` function-declaration* `}`
+
+template-introducer ::= `with` `(` parameter (`,` parameter)* `)`
+parameter ::= identifier `:` expression
+
+typeclass-function ::= function-declaration | function-definition
+function-declaration ::= function-signature `;`
+function-definition ::= function-signature block
+function-signature ::= `function` identifier `(` (parameter (`,` parameter)*)? `)` (`->` trailing-type)?
+```
+
+A definition of a default instance is itself a declaration and cannot be used as an expression. Other definitions of
+typeclass instances are expressions and can be used as such.
+
+
+```
+default-instance-definition ::= `default` `instance` identifier `(` expression `)`
+    `{` function-definition* `}` `;`
+instance-literal ::= `instance` identifier `(` expression `)` `{` function-definition* `}`
+```
+
+## Semantics
+
+This section uses names from the following piece of code to define semantics and requirements:
+
+```
+let tc = with (T : type) typeclass {        // 1
+    function fndecl() -> int;               // 2
+
+    function fndef(arg : int) -> int {      // 3
+        return arg;
+    }
+};
+
+default instance tc(int) {                  // 4
+    function fndecl() -> int {              // 5
```

Comment:
Instances shouldn't have to (be able to?) explicitly state the type of the function they are implementing, as this is specified by the typeclass which is being implemented, so will be the same for all instances for a particular type.